SoC14 wxWebViewChromium merge

sjlamerton opened the pull request:
First attempt at a merge of the 2014 wxWebViewChromium Summer of Code work. Straight rebase of the Haojian's changes onto wxWidgets master. Conflicting commits were skipped as they related to build files that were later removed. One additional commit to fix the msw build files after their recent changes.

hokein committed:
Integrate wxWebChromium to wxwidgets

Also add webview_chromium usage samples.

hokein committed:
Remove CEF1 specific code since we only focus on CEF3.

hokein committed:
Send WEBVIEW_ERROR event after CEF finishing loading.

Currently, WEBVIEW_ERROR event is sent to client on CEF::OnLoadError(),
which may cause the error message of wxInfoBar disapear.

CEF::OnLoadEnd will be invoked after CEF::OnloadError, if WEBVIEW_ERROR
event get processed before CEF::LoadEnd, the wxInfoBar's message will
disappear since the webpage will get loaded in CEF::LoadEnd.

CEF will return "data:text/html,chromewebdata" when url is invalid.

hokein committed:
Nits: fix code style.

hokein committed:
Specify the link to Debug/Release CEF library.

hokein committed:
Nits: Fix code style.

hokein committed:
Add Building Instructions Document.

hokein committed:
Remove unnecessary subprocess.exe during webview startups.

hokein committed:
Correct building link to CEF library.

hokein committed:
Update readme.md

hokein committed:
Add webview_chromium.bkl to samples.bkl

hokein committed:
Enable wxWebviewChromium dynamic builds.

hokein committed:
Move CEF specific variables to config.bkl to avoid duplicate.

hokein committed:
Add a CEF_INCLUDE_DIR check in webview_chromium sample.

hokein committed:
Remove lib prefix in multilib.bkl

There is no need to add lib prefix, since g++ `-l` option will add it
automatically.

This patch does no effect on windows platform.

hokein committed:
Add CEF link path on Linux.

hokein committed:
Split CEF_LIB_DIR into LIBCEF_PATH and LIBCEF_WRAPPER_PATH.

hokein committed:
Add webview_chromium sample in configure.in

Fix configure can not generate Makefile in samples/webview_chromium directory.

hokein committed:
Correct link path in webview_chromium sample.

hokein committed:
Correct cef link name on Windows.

Linux link option does not require 'lib' prefix, while windows
require.

hokein committed:
Integrate CEF MessageLoop to wxWidgets MessageLoop.

* Enable webview_chromium on Linux, webview_chromium runs normally,
  but browser window is split out of wxWidget window now and we need to
  host it in wxWidgets window later.
* Make cef messageloop processing in wx::ProcessIdle period, this way
works on windows and linux.
* Don't use multiple_message_loop on Windows.

hokein committed:
Fix autoconf generated error in bakefile_gen.

hokein committed:
Fix some macro mistake issues.

hokein committed:
Fix build error on windows.

hokein committed:
Set cef messageloop running inside wxWebviewChromium instead of client
side.

hokein committed:
Set wxWidgets_src/cef as CEF default path.

hokein committed:
Update readme.

hokein committed:
Host CEF browser window in wxWidgets window.

hokein committed:
Enable CEF debug log.

hokein committed:
Remove CEF message loop running to client.

Running CEF message loop inside wxWebviewChromium internal will get
segument fault, while moving to client doesn't.

Reset to the previous client solution to avoid segument falut, since I
don't figure out the segument fault reason.

hokein committed:
Fix argument inconsistency issue.

hokein committed:
Fix windows build error.

hokein committed:
Cleanup: remove unused m_timer.

hokein committed:
Add more building instruction in readme.

hokein committed:
Fix gtk_scrolled_window_add() warning.

Using gtk viewport instead of vbox, to avoid the following waring: "
Gtk: gtk_scrolled_window_add(): cannot add non scrollable widget use
gtk_scrolled_window_add_with_viewport() instead".

hokein committed:
Fix rename issue.

hokein committed:
Add webview xcode project file.

Copy and modify from samples/minimal/minimal_cocoa.xcodeproj

hokein committed:
Use double quotes instead of angle brackets to include wx header files.

hokein committed:
Cleanup: remove StartUpSubprocess() since it no longer used.

hokein committed:
Add launch helper for executing separate processes on mac OS X.

hokein committed:
Upgrade to CEF3.1750.1738.

hokein committed:
Enable webview_chromium building through xcode on mac OS X.

hokein committed:
[Mac] Host cef browser window to wxwidgets windows.

hokein committed:
[Mac] Enable webview_chromium dynamic builds.

hokein committed:
[Mac] Rename webview_chromium helper to xcode project.

hokein committed:
[Mac] Correct dynamic dependent project.

hokein committed:
[Mac] Fix find bar is overrided by webview window issue.

hokein committed:
[Mac] Resize browser window.

hokein committed:
Fix a crash issue when close wxwebview_chromium sample.

On Mac OS X, it will trigger a crash issue when user close the window
(It also appear in other platforms, but the crash doesn't appear.)

The crash is caused by calling `CefShutdown` in wxApp::OnExit(). When
wx application running in wxApp::OnExit function, the webframe has been
deleted, we need to call `CefShutdown` in webframe `Close` period.

hokein committed:
Remove unnecessary code.

hokein committed:
[Win] Add VS2010 wxWebviewChromium sample static build supports.

hokein committed:
[Win] Add VS2010 wxWebviewChromium sample dynamic build supports.

hokein committed:
Update readme.md

Add `libcef_dll_wrapper` building steps for wxWidgets.

hokein committed:
Update Readme.md.

hokein committed:
Add Mac OS X building tutorial.

hokein committed:
Disable sandbox in CEF3.1750.1738.

hokein committed:
Use Connect method instead of EVENT_TABLE to hook WebFrame::OnClose event.

hokein committed:
[Win/Linux] A temporary fix crashing when closing WebFrame.

hokein committed:
Add custom handler support.

hokein committed:
Correct help/doc.zip loading path.

Due to webview sample building directory is different from platforms and
compilers, we need to retrieve the `<wxwidgets-src>/help/` absolute path
from building directory.

hokein committed:
Nits: Correct code style.

hokein committed:
[Mac] Add MimeType empty check.

wxFileSystemHandler::GetMimeTypeFromExt returns empty when
passing css location url on Mac OS X, which will trigger DCHECK failure in CEF.

This patch is adding empty check to avoid DCHECK failure.

hokein committed:
[Mac] Fix a warning: '_' is not allowed in bundle file.

hokein committed:
Nits: Fix some compile warnings.

hokein committed:
Update readme: Upgrade to CEF3.1750.1738.

hokein committed:
Rename RunCEFMessageLoopOnIdle to DoCEFWork for better understanding.

hokein committed:
Running CEFMessageLoop in wxTimer instead of OnIdle.

Running CEFMessageLoop form OnIdle, which may pontentially starve CEF
messageloop when application is busy with other UI-operation.

Change it to timer to avoid the starvation.

hokein committed:
Restore OnExit method.

Althrough there is fine to exit the program without calling CEFShutdown on Win/Linux platforms, it's better to
call it explicitly.

hokein committed:
Cleanup some commented code.

hokein committed:
[Mac] Add CEF path to bakefile system.

hokein committed:
Nits: Correct code indent.

hokein committed:
[Mac] Add CEF framework link in configure.in.

hokein committed:
[Mac] Add building scripts from CEF.

hokein committed:
[Mac] Support Makefile builds.

hokein committed:
Add wxUSE_WEBVIEW_CHROMIUM macro.

hokein committed:
Add --enable-webviewchromium option in configure.

hokein committed:
Specify CEF Framework link in bakefile system instead configure.

Add CEF framework link in configure will cause each binary builds with
CEF framework path option. We only need to link CEF framework on
building webview library.

hokein committed:
Add wxUSE_WEBVIEW_CHROMIUM macro check in webview_chromium.h/.cpp.

hokein committed:
Disable webview Chromium feature by Default.

hokein committed:
Set wxUSE_WEBVIEW_CHROMIUM marco in configure.in.

hokein committed:
Use wxUSE_WEBVIEW_CHROMIUM marco check in sample.

hokein committed:
Update readme.

hokein committed:
[Mac] Only build webview_chromium helper on OS X.

hokein committed:
Fix bakefile system generation error, only build helper on OS X.

hokein committed:
Create an independent webview_chromium library.

hokein committed:
Use new webview_chromium library instead of webview.

hokein committed:
Remove chromium related code from webview library.

hokein committed:
[Win] Add Visual Studio 2010 webview_chromium library project files.

hokein committed:
Move cef directory from <wxwidget_dir>/cef to <wxwidget_dir>/src/cef.

hokein committed:
[Linux] Fix bakefile generation error of webview_chromium sample.

hokein committed:
Add USE_WEBVIEW_CHROMIUM option in bkl file to allow build/unbuild
webview_chromium library.

hokein committed:
[Mac] Change webview_chromium sample to use webview_chromium library.

hokein committed:
Fix bakefile generation error on win/linux platforms.

hokein committed:
[Win] Add Visual Studio 2012/2013 webviewchromium library project files.

hokein committed:
[Win] Update webviewchromium Visual Studio project files.

hokein committed:
Add CEF check in configure.

hokein committed:
[Linux] Add libcef.so path in webview_chromium sample no need to specify
in LD_LIBRARY_PATH.

hokein committed:
[Mac] Move mac related resource files to mac directory.

hokein committed:
Add wxWebViewChromium interface document.

hokein committed:
Add wxWebViewFactoryChromium docs and API differences section.

hokein committed:
Remove cef directory check in configure.in.

hokein committed:
Avoid compiler's warning of unused parameters.

hokein committed:
Remove cef lower version support.

Only supports 3.1650.1562(Win/Linux) and 3.1750.1738(Win/Linux/Mac).

hokein committed:
[Win] Correct webview_chromium sample cef dir path.

hokein committed:
[Mac] Enable static builds.

hokein committed:
Update docs: Add required libraries on Linux.

hokein committed:
[Mac] Remove webview sample xcode project.

hokein committed:
[Win] Remove wxUSE_WEBVIEW_CHROMIUM in VS project files to avoid macro
redefinition.

hokein committed:
Update docs.

hokein committed:
[Mac] Make libwebview_chromium depends on libwebview.

This will avoid include webview common code in both two library, and
allow end user to link both two libs for using two backends at the same
times.

hokein committed:
Update docs.

hokein committed:
[Windows] Update to use the new webviewchromium library.

* Update VS2010 project file.
* Update webview_chromium sample VS2010 project file.
* Introduce a WXDLLIMPEXP_WEBVIEW_CHROMIUM and WXDLLIMPEXP_DATA_WEBVIEW_CHROMIUM
  macro to enable dynamic builds on windows.

hokein committed:
Remove wxUSE_WEBVIEW_CHROMIUM check in webview.cpp

hokein committed:
[Win] Fix Debug build error in VS2010.

hokein committed:
[Windows] Update VS2012&2013 project files.

hokein committed:
[Win] Fix wxWebViewChromium sample debug build error in VS2010.

hokein committed:
[Win] Add VS2012&VS2013 build supports for wxWebViewChromium sample.

sjlamerton committed:
Update wxWebViewChromium MSW build files.

hokein committed:
Update Vistual Studio project files.

Remove wx_vc1[012]_webviewchromium files, and use wx_vc_webviewchromium.vcxproj for VS2010/2012/2013 which
follows the trunk.

hokein committed:
Remove wxWebViewChromium sample xcode project file.

sjlamerton committed:
Tidy up wxWebViewChromium documentation.

sjlamerton committed:
Build system fixes for wxWebViewChromium.

sjlamerton committed:
Remove old wxWebViewChromium readme.

sjlamerton committed:
Documentation updates for wxWebViewChromium.

steve-lamerton commented:
@vadz I rebased this earlier today with a few changes since the last time we discussed this, any chance you could take a look again?

sjlamerton committed:
Add support for CEF 1916.

sjlamerton committed:
Add support for CEF 2062.

sjlamerton committed:
Update wxWebViewChromium documentation.

- Add supported CEF versions.
- Clarify required runtime files.

sjlamerton committed:
Increase speed of wxWebViewChromium timer.

vadz commented on configure.in:
```diff
@@ -8233,7 +8239,7 @@ echo "                                       zlib               ${wxUSE_ZLIB}"
 echo "                                       libmspack          ${wxUSE_LIBMSPACK}"
 echo "                                       sdl                ${wxUSE_LIBSDL}"
-
+echo "                                       webview_chromium   ${wxUSE_WEBVIEW_CHROMIUM}"
```
Ideal would be to have something like

    webview         yes (with the following backends: WebKit Chromium)

to provide maximum of information on a single line.

vadz commented on configure.in:
```diff
@@ -7047,6 +7048,11 @@ if test "$wxUSE_WEBVIEW" = "yes"; then
     AC_DEFINE(wxUSE_WEBVIEW)
     SAMPLES_SUBDIRS="$SAMPLES_SUBDIRS webview"
+    if test "$wxUSE_WEBVIEW_CHROMIUM" = "yes"; then
+        USE_WEBVIEW_CHROMIUM=1
```
This doesn't seem right, the option is "yes" by default (not sure if it should be, if not, `DEFAULT_wxUSE_WEBVIEW_CHROMIUM` should be set to "no") and we don't check that we have the required headers/libraries here, so it looks like the compilation will fail out of the box.

We should have `AC_TRY_LINK` check here.

vadz commented on include/wx/webview_chromium.h:
```diff
@@ -0,0 +1,323 @@
+#pragma warning(disable:4100)
+#endif
+
+#include "include/cef_browser.h"
```
This would probably represent a significant change, but it would be great if we could avoid including all this CEF stuff from this public header. AFAICS the only thing actually needed here is `m_clientHandler` and we could just wrap it into some `wxWebViewChromiumImpl` to avoid having it here.

There would be quite a few advantages of keeping this header free of CEF stuff, in particular we could even allow people to use this backend without installing CEF headers at all if we just build the official binaries with it, it would be enough to distribute just the DLLs.

vadz commented on interface/wx/webview_chromium.h:
```diff
@@ -0,0 +1,256 @@
+
+    __General__
+
+    The wxWebViewChromium backend is built into a seperate webview_chromium
```
Typo s/seperate/separate/

vadz commented on interface/wx/webview_chromium.h:
```diff
@@ -0,0 +1,256 @@
+    library which depends on the webview library.
+
+    Once you have a copy of CEF3, either from compiling it yourself or using
+    prebuild binaries you should copy it into `wx_root/src/cef`. To run the
```
s/prebuild/prebuilt/

vadz commented on samples/webview/Info_cocoa.plist:
```diff
@@ -0,0 +1,41 @@
+		<string>it</string>
+	</array>
+	<key>CFBundleLongVersionString</key>
+	<string>3.1.0, (c) 2005-2013 wxWidgets</string>
```
Very minor, but copyright year is off by 2 by now...

(here and below)

vadz commented on src/common/webview_chromium.cpp:
```diff
@@ -0,0 +1,809 @@
+class wxStringVisitor : public CefStringVisitor
+{
+public:
+    enum StringType {
```
Minor: `{` should be on its own line.

vadz commented on src/common/webview_chromium.cpp:
```diff
@@ -0,0 +1,809 @@
+           long style,
+           const wxString& name)
+{
+#ifdef __WXMSW__
```
What's the difference between the MSW and OSX case? I.e. why don't we check for the return value of `Create()` in the latter and why don't we just have `#ifdef __WXGTK__` and `#else` branches instead of checking for each platform?

vadz commented on src/common/webview_chromium.cpp:
```diff
@@ -0,0 +1,809 @@
+    }
+#endif
+
+#ifdef __WXGTK__
```
A comment explaining why is this needed and, in particular, why can't we just call base class `Create()` with `style | wxHSCROLL | wxVSCROLL` would be helpful.

vadz commented on src/common/webview_chromium.cpp:
```diff
@@ -0,0 +1,809 @@
+    CefMainArgs args(argc, argv);
+#endif
+    // If there is no subprocess then we need to execute on this process
+    if ( path == "" )
```
Minor: should be `path.empty()`

vadz commented on src/common/webview_chromium.cpp:
```diff
@@ -0,0 +1,809 @@
+    settings.no_sandbox = true;
+#endif
+
+#ifdef __WXDEBUG__
```
This is not appropriate, `__WXDEBUG__` is defined even in release build by default in wx3.

It would probably be better to either use some `wxWEBVIEW_DEBUG` environment variable or provide a method for enabling debug logging.

vadz commented:
Thanks again for your work, I can't really comment on the interesting parts, i.e. the code actually using CEF API, but globally it seems good except for the few things pointed out in the comments.

I also wonder why do we need a separate sample for this, could we just allow using this backend as an option in the normal webview sample? I'd really like to have just one sample instead of two similar but different ones.

sjlamerton committed:
Typo fixes in wxWebViewChromium docs.

sjlamerton committed:
Update wxWebViewChromium copyright dates.

sjlamerton committed:
wxWebViewChromium style fixes.

sjlamerton committed:
List wxWebView backends in configure output

sjlamerton committed:
Improve wxWebView configure script backend searching

- Only disable wxWebView if no backends are present
- Check for Chromium headers if enabled
- Remove some unused variables

steve-lamerton commented:
I have made most of the fixes, the Impl changes would be quite big but I'll take a look later this week, I'll also update the debug output then. As for the samples I think the reason that there are two is because it isn't possible to compile the webview_chromium library on Windows if CEF is not present which would limit the number of people who could build the sample as it would fail to link.

vadz commented:
About the sample: could we at least reuse the sources? I.e. we could still have a different directory with its own makefiles, but just `#include "../webview/webview.cpp"` from `webview_chromium.cpp` or something like this. Duplicating so much code really doesn't sit well with me.

mths0x5f commented:
Sorry for the intromission, but how is this going? What can someone do to help?

steve-lamerton commented:
The only item left to resolve before the merge was the duplicated code in the samples, it wouldn't take much to resolve that and then I think we should be able to close this pull request.

mths0x5f commented:
Well... I do not know if it helps, but I have edited some code and made a pull request to you, @steve-lamerton. I'm sorry if inconvenient.

https://github.com/steve-lamerton/wxWidgets/pull/1

steve-lamerton commented:
Thanks, I will take a look later this week.

tierra commented:
It's probably too late to look into getting this into 3.1.0, but it would still be great if @sjlamerton or @hokein could look at rebasing this, resolving conflicts, and testing this out on latest master. I'd still love to see this make it into a future 3.1.x release even if it's not 3.1.0.

mtangoo commented:
I would also love to see this merged. It will relieve developers relying on web a lot of pain especially with versions of Native engines available (some are still using very old MS browsers in their systems)

vadz commented:
I believe this is obsoleted by #706, please let me know if I was wrong to close this one.
